src/main/java/com/google/research/bleth/simulator/Location.java

```diff
@@ -1,40 +1,23 @@
 package com.google.research.bleth.simulator;
 
-import java.util.Objects;
+import com.google.auto.value.AutoValue;
 
 /** A location on board. */
-public class Location {
-    public final int row;
-    public final int col;
-
-    public Location(int row, int col) {
-        this.row = row;
-        this.col = col;
-    }
-
-    @Override
-    public boolean equals(Object o) {
-        if (this == o) {
-            return true;
-        }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
-        }
-        Location location = (Location) o;
-        return row == location.row && col == location.col;
+@AutoValue
+public abstract class Location {
+    public static Location create(int row, int col) {
+        return new AutoValue_Location(row, col);
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(row, col);
-    }
+    public abstract int row();
+    public abstract int col();
 
     /**
      * Calculate and return the new location based on the current location and the given direction.
      * @param direction is the direction to move to it, one of the four directions: up, down, right and left.
      * @return new Location, differ from the original location in one step, according the given direction.
      */
     public Location moveInDirection(Direction direction) {
-        return new Location(row + direction.getRowDelta(), col + direction.getColDelta());
+        return Location.create(row() + direction.getRowDelta(), col() + direction.getColDelta());
```

Comment:
Nit: no need for `Location.create` here (you can just call the method `create`, as it's in the current scope).